lib/mongodb.nit

```diff
@@ -0,0 +1,708 @@
+	`}
+
+	# Checks if a collection named `name` exists.
+	#
+	# ~~~
+	# var client = new MongoClient("mongodb://localhost:27017/")
+	# var db = client.database("test")
+	# assert not db.has_collection("qwerty")
+	# client.close
+	# ~~~
+	fun has_collection(name: String): Bool import set_mongoc_error, String.to_cstring `{
+		bson_error_t error;
+		if(!mongoc_database_has_collection(recv, String_to_cstring(name), &error)) {
+			MongoDb_set_mongoc_error(recv, &error);
+			return false;
+		}
+		return true;
+	`}
+
+	# Drop `self`, returns false is an error occured.
+	fun drop: Bool import set_mongoc_error, destroy `{
+		bson_error_t error;
+		if(!mongoc_database_drop(recv, &error)) {
+			MongoDb_set_mongoc_error(recv, &error);
+			return false;
+		}
+		MongoDb_destroy(recv);
+		return true;
+	`}
+
+	# FIXME: how to handle destructors?
+	fun destroy `{ mongoc_database_destroy(recv); `}
+
+	# Utility method to set `Sys.last_mongoc_error`.
+	private fun set_mongoc_error(err: BSONError) do sys.last_mongoc_error = err
+end
+
+# A Mongo collection.
+#
+# Collections are automatically created on the MongoDB server upon insertion of
+# the first document.
+# There is no need to create a database manually.
+extern class MongoCollection `{ mongoc_collection_t * `}
+	# Loads a collection.
+	#
+	# Call `MongoDb::collection` instead.
+	private new(client: MongoClient, db, collection: String) import String.to_cstring `{
+		return mongoc_client_get_collection (
+			client,
+			String_to_cstring(db),
+			String_to_cstring(collection));
+	`}
+
+	# Gets the collection name.
+	#
+	# ~~~
+	# var client = new MongoClient("mongodb://localhost:27017/")
+	# var db = client.database("test")
+	# assert db.collection("test").name == "test"
+	# client.close
+	# ~~~
+	fun name: String import NativeString.to_s `{
+		const char *str = mongoc_collection_get_name(recv);
+		return NativeString_to_s((char*)str);
+	`}
+
+	# Inserts a new document in the collection.
+	#
+	# If no _id element is found in document, then a new one be generated locally
+	# and added to the document.
+	#
+	# Returns `false` if an error occured. See `Sys::last_mongoc_error`.
+	#
+	# ~~~
+	# var client = new MongoClient("mongodb://localhost:27017/")
+	# var col = client.database("test").collection("test")
+	# var doc = new JsonObject
+	# doc["foo"] = 10
+	# doc["bar"] = "bar"
+	# doc["baz"] = new JsonArray
+	# assert col.insert(doc)
+	# client.close
+	# ~~~
+	fun insert(doc: JsonObject): Bool
+		import JsonObject.to_bson, set_mongoc_error `{
+		bson_t *bson = JsonObject_to_bson(doc);
+		bson_error_t error;
+		if(!mongoc_collection_insert(recv, MONGOC_INSERT_NONE, bson, NULL, &error)) {
+			MongoDb_set_mongoc_error(recv, &error);
+			return false;
+		}
+		return true;
+	`}
+
+	# Inserts multiple documents in the collection.
+	fun insert_all(docs: Collection[JsonObject]): Bool do
+		var res = true
+		for doc in docs do res = insert(doc)
+		return res
```

Comment:
What is the return? Success or error? Now it only returns the last result, the for loop should probably be:

~~~
for doc in docs do res = insert(doc) and res
~~~